### ISSUE TITLE
Version conflicts for web-streams-polyfill

### ISSUE THREAD

**Issue by yacchin1205** [yacchin1205/ep_kodama#1]
Commenting out the following Workaround code will result in an error at runtime.

https://github.com/yacchin1205/ep_kodama/blob/5d9944bf8d8fb1cbecad4333790cebde71226277/Dockerfile#L25-L26

The details of the error are as follows:

```
etherpad-1  | [2024-04-06T22:49:19.907] [ERROR] settings - [ep_kodama] Error: Cannot find module '/opt/etherpad-lite/src/node_modules/web-streams-polyfill/dist/ponyfill.es2018.js'
etherpad-1  |     at createEsmNotFoundErr (node:internal/modules/cjs/loader:1179:15)
etherpad-1  |     at finalizeEsmResolution (node:internal/modules/cjs/loader:1167:15)
etherpad-1  |     at resolveExports (node:internal/modules/cjs/loader:589:14)
etherpad-1  |     at Module._findPath (node:internal/modules/cjs/loader:666:31)
etherpad-1  |     at Module._resolveFilename (node:internal/modules/cjs/loader:1128:27)
etherpad-1  |     at a._resolveFilename (/opt/etherpad-lite/node_modules/.pnpm/tsx@4.7.1/node_modules/tsx/dist/cjs/index.cjs:1:1729)
etherpad-1  |     at Module._load (node:internal/modules/cjs/loader:983:27)
etherpad-1  |     at Module.require (node:internal/modules/cjs/loader:1230:19)
etherpad-1  |     at require (node:internal/modules/helpers:179:18)
etherpad-1  |     at Object.<anonymous> (/opt/etherpad-lite/src/plugin_packages/openai/_shims/node-runtime.js:42:30)
etherpad-1  |     at Module._compile (node:internal/modules/cjs/loader:1368:14)
etherpad-1  |     at Object.S (/opt/etherpad-lite/node_modules/.pnpm/tsx@4.7.1/node_modules/tsx/dist/cjs/index.cjs:1:1292)
etherpad-1  |     at Module.load (node:internal/modules/cjs/loader:1205:32)
etherpad-1  |     at Module._load (node:internal/modules/cjs/loader:1021:12)
etherpad-1  |     at Module.require (node:internal/modules/cjs/loader:1230:19)
etherpad-1  |     at require (node:internal/modules/helpers:179:18) {
etherpad-1  |   code: 'MODULE_NOT_FOUND',
etherpad-1  |   path: '/opt/etherpad-lite/src/node_modules/web-streams-polyfill/package.json'
etherpad-1  | }
```

According to the web-streams-polyfill documentation https://github.com/MattiasBuelens/web-streams-polyfill/blob/master/MIGRATING.md, ponyfill.es2018.js import method seems to have changed in v4. This may be due to version conflicts.

**Comment by yacchin1205** [comment]
I checked the version of web-streams-polyfill installed after installing this plugin and it is `4.0.0-beta.3`.

I assume the error occurs because openai-node requires `^3.2.1` as Version Range (`3.3.3` in latest npmjs).

https://github.com/openai/openai-node/blob/018ac718ccf6a96798ef8f91906b3b652aa50919/package.json#L35

```
$ cat /opt/etherpad-lite/src/plugin_packages/web-streams-polyfill/package.json 
{
  "name": "web-streams-polyfill",
  "version": "4.0.0-beta.3",
  "description": "Web Streams, based on the WHATWG spec reference implementation",
  "main": "dist/ponyfill.js",
  "module": "dist/ponyfill.mjs",
  "types": "types/ponyfill.d.ts",
```

**Comment by yacchin1205** [comment]
formdata-node requested web-streams-polyfill `4.0.0-beta.3`. This may be why 4.0.0-beta.3 was installed. Doesn't this happen in other environments?

https://github.com/openai/openai-node/blob/018ac718ccf6a96798ef8f91906b3b652aa50919/yarn.lock#L1724

**Comment by yacchin1205** [comment]
I thought it might be a problem when running in pnpm, so I created a verification code, but openai-node worked fine in both npm and pnpm.

https://github.com/yacchin1205/openai-node-version-conflict

The problem may only occur when running on the Etherpad-lite plugin system.

**Comment by Sblop** [comment]
I simply love this plugin, but I do have some problems getting it to work reliably... 

I use the prebuild docker container of etherpad (etherpad/etherpad:latest). I have modified the command section in my docker compose to this:
```
    command: >
      sh -c "
      pnpm config set store-dir /opt/etherpad-lite/src ;
      pnpm config set ignore-workspace-root-check true ;
      pnpm i ep_kodama -w ;
      pnpm i sharp@latest -w ;
      pnpm i openai -w ;
      pnpm i web-streams-polyfill@3.3.3 -w ;
      exec pnpm run prod --apikey REDACTED"
```

I however keep getting errors that polyfill is not found...

I was trying to run commands using docker exec to manually do some of these installations, and at some time yesterday I got it working, which were great. Now, however it keeps failing. What am I doing wrong?

Is it understood correctly, that I do not need to build "your" docker image, but can use the existing image, and simply install ep_kodema?

**Comment by yacchin1205** [comment]
> I simply love this plugin, but I do have some problems getting it to work reliably...

@Sblop Thanks for the comment.
I checked how to build using the existing image. You can install in the following steps.

```
    command: >
      sh -c "
      pnpm install-plugins ep_kodama &&
      cd ./src/ &&
      pnpm i web-streams-polyfill@3.3.3 &&
      cd .. &&
      exec pnpm run prod --apikey REDACTED"    
```

- I think ep_kodama should be installed with `pnpm install-plugins`. (This will change to `pnpm plugins install` in the near future.)
- `web-streams-polyfill@3.3.3` should be installed under the `/opt/etherpad-lite/src` directory.

I have put the file on gist. https://gist.github.com/yacchin1205/125ce26dcb8771237982fbc01192957f

In this example, there must be a `settings.json` in the directory where the `ep_kodama` settings are defined.
https://github.com/yacchin1205/ep_kodama?tab=readme-ov-file#configure-the-openai-api-settings-in-the-settingsjson-file

**Comment by yacchin1205** [comment]
Resolved by #3
